### PR TITLE
Vectorize or scalarize VectorSupport.broadcastCoerced

### PR DESCRIPTION
**Pull request by gita-omr**
- handle VectorSupport.broadcastCoerced during VectorAPIExpansion
- remove checkcast and NULLCHK code since it seems to be incorrect
- replace all int types with int32_t
- add missing /return descriptions to the doxygen comments

### CODE REVIEW

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp [hunk 1]
```diff
@@ -1347,6 +1356,80 @@ TR::Node *TR_VectorAPIExpansion::binaryIntrinsicHandler(TR_VectorAPIExpansion *o
...
+      traceMsg(comp, "broadcastCoercedIntrinsicHandler for node %p\n", node);
+
+   int32_t elementSize = OMR::DataType::getSize(elementType);
+   TR::Node *valueToBroadcast = node->getChild(3);
+
+   anchorOldChildren(opt, treeTop, node);
+
+   TR::Node *newNode;
+
+   switch (elementType) {
```
Comment (jdmpapin) [inline runtime/compiler/optimizer/VectorAPIExpansion.cpp:1386]: Is this switch supposed to be exhaustive? It looks like it, because if none of the cases run, then `newNode` will be left uninitialized. If that's right, then it's probably a good idea to add a default case that fails an assertion

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp [hunk 2]
```diff
@@ -1347,6 +1356,80 @@ TR::Node *TR_VectorAPIExpansion::binaryIntrinsicHandler(TR_VectorAPIExpansion *o
...
+          newNode = TR::Node::create(node, TR::l2b, 1, valueToBroadcast);
+          break;
+      case TR::Int16:
+          newNode = TR::Node::create(node, TR::l2s, 1, valueToBroadcast);
+          break;
+      case TR::Int32:
+          newNode = TR::Node::create(node, TR::l2i, 1, valueToBroadcast);
+          break;
+      case TR::Int64:
+          newNode = TR::Node::create(node, TR::dbits2l, 1, TR::Node::create(node, TR::lbits2d, 1, valueToBroadcast));
```
Comment (jdmpapin) [inline runtime/compiler/optimizer/VectorAPIExpansion.cpp:1403]: I suppose these redundant conversions are to make sure that the `doScalarization` path can easily recreate `node` as a copy of `newNode`. I would recommend `PassThrough` for that purpose instead, but it looks like its `getDataType()` will return `NoType`, which I don't think will work with vector opcode type inference - maybe we should change that

Anyway, it might be good to have a code comment to explain what's going on here

**Comment by jdmpapin** [comment]
Also, in the commit message and initial PR comment: "handle NULLCHK and PassThrough nodes since [... what?]"

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp (hunk 2 above)
Comment (gita-omr) [reply runtime/compiler/optimizer/VectorAPIExpansion.cpp:1403]: Yes, that's correct: just to make node recreation easier. I will put a comment of course.

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp [hunk 3]
```diff
@@ -350,10 +350,6 @@ TR_VectorAPIExpansion::visitNodeToBuildVectorAliases(TR::Node *node)
          invalidateSymRef(child->getSymbolReference());
          }
       }
-   else if (node->getOpCodeValue() == TR::checkcast)
-      {
-      // ignore for now and check the children
-      }
```
Comment (jdmpapin) [inline runtime/compiler/optimizer/VectorAPIExpansion.cpp:356]: Was there a problem with `checkcast` here as well?

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp [hunk 4]
```diff
@@ -793,19 +795,6 @@ TR_VectorAPIExpansion::expandVectorAPI()
          TR::Node::recreate(parent, TR::treetop);
          methodTable[handlerIndex]._methodHandler(this, treeTop, node, elementType, vectorLength, doMode);
          }
-      else if (node->getOpCodeValue() == TR::checkcast)
```
Comment (jdmpapin) [inline runtime/compiler/optimizer/VectorAPIExpansion.cpp:796]: I take it that you agree that this code was dead :sweat_smile: 

The `checkcast`-related deletion(s) seem to be unrelated to the other changes in this commit though, so I think they warrant a separate commit or at least a separate mention in the commit message

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp (hunk 3 above)
Comment (gita-omr) [reply runtime/compiler/optimizer/VectorAPIExpansion.cpp:356]: yes, if the code during transformation is dead it's better not to ignore both checkcast and NULLCHK during the analysis. I will add a comment to this PR and will open a separate PR for those.

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp (hunk 1 above)
Comment (gita-omr) [reply runtime/compiler/optimizer/VectorAPIExpansion.cpp:1368]: These are the only element types that are supported by the library. But I can add and assert.

Comment (gita-omr) [reply runtime/compiler/optimizer/VectorAPIExpansion.cpp:1368]: done

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp (hunk 2 above)
Comment (gita-omr) [reply runtime/compiler/optimizer/VectorAPIExpansion.cpp:1386]: added a comment

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp (hunk 3 above)
Comment (gita-omr) [reply runtime/compiler/optimizer/VectorAPIExpansion.cpp:356]: done

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp (hunk 4 above)
Comment (gita-omr) [reply runtime/compiler/optimizer/VectorAPIExpansion.cpp:796]: yeah, added to the commit message.

**Review by jdmpapin — APPROVED** [review]

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp [hunk 5]
```diff
@@ -369,6 +365,12 @@ TR_VectorAPIExpansion::visitNodeToBuildVectorAliases(TR::Node *node)
          }
       }
 
+   // skip PassTrough only if it's a child of a known parent
```
Comment (0xdaryl) [inline runtime/compiler/optimizer/VectorAPIExpansion.cpp:368]: sp.  PassThrough

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp [hunk 6]
```diff
@@ -1351,6 +1338,84 @@ TR::Node *TR_VectorAPIExpansion::binaryIntrinsicHandler(TR_VectorAPIExpansion *o
...
+          break;
+      case TR::Int32:
+          newNode = TR::Node::create(node, TR::l2i, 1, valueToBroadcast);
+          break;
+      case TR::Int64:
+          // redundant conversion to simplify node recreation
+          newNode = TR::Node::create(node, TR::dbits2l, 1, TR::Node::create(node, TR::lbits2d, 1, valueToBroadcast));
+          break;
+      default:
+         TR_ASSERT_FATAL(false, "Was a new element type added to VectorAPI?\n");
```
Comment (0xdaryl) [inline runtime/compiler/optimizer/VectorAPIExpansion.cpp:1389]: Rather than ask a question, just say the element type was unexpected and ideally reporting the type received.

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp [hunk 7]
```diff
@@ -1351,6 +1338,84 @@ TR::Node *TR_VectorAPIExpansion::binaryIntrinsicHandler(TR_VectorAPIExpansion *o
...
+         break;
+   }
+
+   if (mode == doScalarization)
+      {
+      // modify original node in place
+      node->setAndIncChild(0, newNode->getChild(0));
+      node->setNumChildren(1);
+
+      int numLanes = vectorLength/8/elementSize;
```
Comment (0xdaryl) [inline runtime/compiler/optimizer/VectorAPIExpansion.cpp:1399]: inconsistent use of `int` / `int32_t` in this method.  The code base preference is `int32_t`.  Another instance in the loop below.

File: runtime/compiler/optimizer/VectorAPIExpansion.hpp [hunk 8]
```diff
@@ -665,6 +665,32 @@ class TR_VectorAPIExpansion : public TR::Optimization
...
+   *
+   *   \param elementType
+   *      Element type
+   *
+   *   \param vectorLength
+   *      Full vector length in bits (e.g. 128 for Float128Vector)
+   *
+   *   \param mode
+   *      Handler mode
+   *
```
Comment (0xdaryl) [inline runtime/compiler/optimizer/VectorAPIExpansion.hpp:690]: Missing `\returns` describing return value.

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp (hunk 7 above)
Comment (gita-omr) [reply runtime/compiler/optimizer/VectorAPIExpansion.cpp:1399]: Sorry, not sure what you mean by the loop below. Is it the type if "i" ?

Comment (gita-omr) [reply runtime/compiler/optimizer/VectorAPIExpansion.cpp]: Replaced all ints with int32_t. Added comment to the commit message.

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp (hunk 6 above)
Comment (gita-omr) [reply runtime/compiler/optimizer/VectorAPIExpansion.cpp]: Changed the message.

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk17

**Comment by 0xdaryl** [comment]
See the Power Linux crash during the build.  It doesn't seem likely to be related to your code, but please investigate nonetheless.

**Comment by gita-omr** [comment]
> 
> 
> See the Power Linux crash during the build. It doesn't seem likely to be related to your code, but please investigate nonetheless.

It's unrelated to my change and I opened an issue: https://github.com/eclipse-openj9/openj9/issues/13897

**Comment by 0xdaryl** [comment]
Jenkins test sanity plinux jdk17

**Review by 0xdaryl — APPROVED** [review]
